src/pybind/mgr/object_format.py

```diff
@@ -0,0 +1,487 @@
+        self.format_name = format_name
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return -errno.EINVAL, "", f"Unknown format name: {self.format_name}"
+
+
+class UnsupportedFormat(ErrorResponseBase):
+    """Raised if the format name does not correspond to any valid
+    conversion functions.
+    """
+
+    def __init__(self, format_name: str) -> None:
+        self.format_name = format_name
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return -errno.EINVAL, "", f"Unsupported format: {self.format_name}"
+
+
+class ErrorResponse(ErrorResponseBase):
+    """General exception convertible to a mgr response."""
+
+    E = TypeVar("E", bound="ErrorResponse")
+
+    def __init__(self, status: str, return_value: Optional[int] = None) -> None:
+        self.return_value = (
+            return_value if return_value is not None else -errno.EINVAL
+        )
+        self.status = status
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return (self.return_value, "", self.status)
+
+    def mgr_return_value(self) -> int:
+        return self.return_value
+
+    @property
+    def errno(self) -> int:
+        rv = self.return_value
+        return -rv if rv < 0 else rv
+
+    def __repr__(self) -> str:
+        return f"ErrorResponse({self.status!r}, {self.return_value!r})"
+
+    @classmethod
+    def wrap(
+        cls: Type[E], exc: Exception, return_value: Optional[int] = None
+    ) -> E:
+        if return_value is None:
+            try:
+                return_value = -int(getattr(exc, "errno"))
+            except (AttributeError, ValueError):
+                pass
+        err = cls(str(exc), return_value=return_value)
+        setattr(err, "__cause__", exc)
+        return err
+
+
+def _get_requested_format(f: ObjectResponseFuncType, kw: Dict[str, Any]) -> str:
+    # todo: leave 'format' in kw dict iff its part of f's signature
+    return kw.pop("format", None)
+
+
+class Responder:
+    """A decorator type intended to assist in converting Python return types
+    into valid responses for the Ceph MGR.
+
+    A function that returns a Python object will have the object converted into
+    a return value and formatted response body, based on the `format` argument
+    passed to the mgr.
+
+    The Responder object is callable and is expected to be used as a decorator.
+    """
+
+    def __init__(
+        self, formatter: Optional[Callable[..., CommonFormatter]] = None
+    ) -> None:
+        self.formatter = formatter
+        self.default_format = "json"
+
+    def _formatter(self, obj: Any) -> CommonFormatter:
+        """Return the formatter/format-adapter for the object."""
+        if self.formatter is not None:
+            return self.formatter(obj)
+        return ObjectFormatAdapter(obj)
+
+    def _retval_provider(self, obj: Any) -> ReturnValueProvider:
+        """Return a ReturnValueProvider for the given object."""
+        return ReturnValueAdapter(obj)
```

Comment:
Unless the `ReturnValueAdapter` is a nested class of this class (it could be, since no one else is calling it), this should be `staticmethod`, shouldn't it?

Reply:
I wanted it to be able to make use of self from a subclass, in case we need to subclass it in the future.